### PR TITLE
RevEng: Use ConfigurationFactory to create all *Configuration objects.

### PR DESCRIPTION
**Pull request by lajones**
@natemcmaster Would like to check this in before the API review tomorrow. Wanted to update so that we use the ConfigurationFactory to create all *Configuration objects rather than just ModelConfiguration. This allows providers to override with new classes to provide additional info if needed.

### CODE REVIEW

File: src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringGenerator.cs [hunk 1]
```diff
@@ -18,7 +18,7 @@ namespace Microsoft.Data.Entity.Relational.Design.ReverseEngineering
 {
     public class ReverseEngineeringGenerator
     {
-        private readonly ModelConfigurationFactory _modelConfigurationFactory;
+        private readonly ConfigurationFactory _modelConfigurationFactory;
```
Comment (smitpatel) [inline src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringGenerator.cs]: nit: change variable name too.

File: src/EntityFramework.SqlServer.Design/ReverseEngineering/SqlServerModelConfigurationFactory.cs [hunk 2]
```diff
@@ -10,7 +10,7 @@
 
 namespace Microsoft.Data.Entity.SqlServer.Design.ReverseEngineering
 {
-    public class SqlServerModelConfigurationFactory : ModelConfigurationFactory
+    public class SqlServerModelConfigurationFactory : ConfigurationFactory
```
Comment (smitpatel) [inline src/EntityFramework.SqlServer.Design/ReverseEngineering/SqlServerModelConfigurationFactory.cs]: `SqlServerConfigurationFactory`? (and likewise for sqlite)

File: src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringGenerator.cs (hunk 1 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/ReverseEngineeringGenerator.cs]: Agreed - I'll check throughout.

File: src/EntityFramework.Relational.Design/ReverseEngineering/Configuration/ModelConfiguration.cs [hunk 3]
```diff
@@ -192,15 +199,13 @@ public virtual void AddTableNameConfiguration([NotNull] EntityConfiguration enti
                 var delimitedSchemaName =
                     CSharpUtilities.DelimitString(ExtensionsProvider.For(entityType).Schema);
                 entityConfiguration.FluentApiConfigurations.Add(
-                    new FluentApiConfiguration(
+                    _configurationFactory.CreateFluentApiConfiguration(
+                        true,
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/ReverseEngineering/Configuration/ModelConfiguration.cs]: Nit: style preference. For boolean literals, use named parameters (will require rearranging). Makes code more readable.

File: src/EntityFramework.SqlServer.Design/ReverseEngineering/SqlServerModelConfigurationFactory.cs (hunk 2 above)
Comment (lajones) [inline src/EntityFramework.SqlServer.Design/ReverseEngineering/SqlServerModelConfigurationFactory.cs]: Agreed - will fix.

File: src/EntityFramework.Relational.Design/ReverseEngineering/Configuration/ModelConfiguration.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/Configuration/ModelConfiguration.cs]: Can't - tried but you can't mix it with params. Happy to add /* hasAttributeEquivalent */ if you like?

Comment (smitpatel) [inline src/EntityFramework.Relational.Design/ReverseEngineering/Configuration/ModelConfiguration.cs]: Last argument to function is params type. Use inline comment to specify parameter name?

File: src/EntityFramework.Relational.Design/ReverseEngineering/Configuration/ModelConfiguration.cs [hunk 4]
```diff
@@ -23,22 +23,26 @@ public abstract class ModelConfiguration
         protected static readonly KeyDiscoveryConvention _keyDiscoveryConvention = new KeyDiscoveryConvention();
         protected static readonly KeyConvention _keyConvention = new KeyConvention();
 
+        protected readonly ConfigurationFactory _configurationFactory;
         protected List<OptionsBuilderConfiguration> _onConfiguringConfigurations;
         protected SortedDictionary<EntityType, EntityConfiguration> _entityConfigurationMap;
 
         public ModelConfiguration(
+            [NotNull] ConfigurationFactory configurationFactory,
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/ReverseEngineering/Configuration/ModelConfiguration.cs]: This is the only quirky part about this change. It just seems odd to give a reference to the factory in a ctor. But don't know there is a better way to get around this...

**Comment by natemcmaster** [comment]
:shipit: (with the minor var names changed...and perhaps the factory reference thing fixed too i possible)

File: src/EntityFramework.Relational.Design/ReverseEngineering/Configuration/ModelConfiguration.cs (hunk 4 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/Configuration/ModelConfiguration.cs]: Agreed. But I couldn't think of a quick way around it. In design mtg would like to discuss using a builder approach. The builder is passed the ConfigurationFactory but the builder is what creates the *Configuration objects. But this is just a quick fix for tomorrow's API review.

**Comment by lajones** [comment]
Checked in with commit cfe32e0.

**Pull request closed without merging**